crates/bevy_ui/src/render/mod.rs

```diff
@@ -522,125 +529,72 @@ pub fn extract_uinode_borders(
         .map(|r| r * ui_scale.0);
 
         let border_radius = clamp_radius(border_radius, uinode.size(), border.into());
-        let transform = global_transform.compute_matrix();
 
-        extracted_uinodes.uinodes.insert(
-            commands.spawn_empty().id(),
-            ExtractedUiNode {
-                stack_index: uinode.stack_index,
-                // This translates the uinode's transform to the center of the current border rectangle
-                transform,
-                color: border_color.0.into(),
-                rect: Rect {
-                    max: uinode.size(),
-                    ..Default::default()
-                },
-                image,
-                atlas_scaling: None,
-                clip: clip.map(|clip| clip.clip),
-                flip_x: false,
-                flip_y: false,
-                camera_entity,
-                border_radius,
-                border,
-                node_type: NodeType::Border,
-            },
-        );
-    }
-}
-
-pub fn extract_uinode_outlines(
-    mut commands: Commands,
-    mut extracted_uinodes: ResMut<ExtractedUiNodes>,
-    default_ui_camera: Extract<DefaultUiCamera>,
-    uinode_query: Extract<
-        Query<(
-            &Node,
-            &GlobalTransform,
-            &ViewVisibility,
-            Option<&CalculatedClip>,
-            Option<&TargetCamera>,
-            &Outline,
-        )>,
-    >,
-) {
-    let image = AssetId::<Image>::default();
-    for (node, global_transform, view_visibility, maybe_clip, camera, outline) in &uinode_query {
-        let Some(camera_entity) = camera.map(TargetCamera::entity).or(default_ui_camera.get())
-        else {
-            continue;
-        };
-
-        // Skip invisible outlines
-        if !view_visibility.get()
-            || outline.color.is_fully_transparent()
-            || node.outline_width == 0.
+        // don't extract border if no border or the node is zero-sized (a zero sized node can have an outline).
+        if uinode.size().x > 0.
+            && uinode.size().y > 0.
+            && !(left == 0.0 && top == 0.0 && right == 0.0 && bottom == 0.0)
```

Comment:
Could you turn this condition into two boolean variables with descriptive name instead of having a comment?

Reply:
The comment here was meant more as a reminder that even a uinode with zero area can still have an outline drawn around it I guess.

Thinking about it there would be no need for the uinode size checks at all actually if the border values were calculated in the layout system instead.

Reply:
Also realized there's another bug, outlines shouldn't be drawn when `Display::None` is set.

Reply:
This is a bit better now with the border array check instead of comparing the individual values. Moving the border calculations to `ui_layout_system` is definitely the way to go as then we can just skip the node size check.

Reply:
Good catch :) I'd still be happier if the condition was split into two nice booleans, maybe the following?
```rust
let has_border = border != [0.; 4];
let node_has_size = uinode.size().x > 0. && uinode.size().y > 0.;
```